// cognitive-complexity-ts/build/ui/ts/component/tree/Folder.js
import { ToggleableBox } from "../box/ToggleableBox.js";
import { CopyText } from "../controls/CopyText.js";
import { concatFilePath } from "../../domain/path.js";
import { showInTree, StickyTitle } from "./StickyTitle.js";
export class Folder {
    constructor(folder, children) {
        this.content = children;
        this.depth = folder.depth;
        const fullPath = concatFilePath(folder.path, folder.name);
        this.title = new StickyTitle([
            folder.name,
            CopyText(fullPath),
        ], folder.depth);
        this.box = new ToggleableBox([
            this.title.dom,
        ], false, () => this.scrollIntoView());
        this.box.changeHideableContent(() => [this.content.dom]);
    }
    get dom() {
        return this.box.dom;
    }
    scrollIntoView() {
        showInTree(this.dom, this.depth);
    }
    setDepth(depth) {
        if (this.depth !== depth) {
            this.depth = depth;
            this.title.setDepth(depth);
        }
    }
    setOpenness(open) {
        this.box.setOpenness(open);
    }
}
